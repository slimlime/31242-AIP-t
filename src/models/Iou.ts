import { Model } from "sequelize";
import Item from "./Item";
import Offer from "./Offer";
import User from "./User";

// Class definition

export interface IIouAttributes {
  id: string;
  item: string;
  giver: string;
  receiver?: string;
  parent_offer?: string;
  proof_of_debt?: string;
  proof_of_completion?: string;
  created_time: Date;
  claimed_time?: Date;
  is_claimed: boolean;
}

class Iou extends Model<IIouAttributes> implements IIouAttributes {
  public id!: string;
  public item!: string;
  public giver!: string;
  public receiver: string | undefined;
  public parent_offer!: string;
  public proof_of_debt: string | undefined;
  public proof_of_completion: string | undefined;
  public created_time!: Date;
  public claimed_time: Date | undefined;
  public is_claimed!: boolean;
}

// IOU relationships

export async function initialiseIouRelationships() {
  const ItemForeignKey = {
    foreignKey: {
      name: "item",
      allowNull: false,
    },
  };
  Iou.belongsTo(Item, ItemForeignKey);
  Item.hasMany(Iou, ItemForeignKey);

  const GiverForeignKey = {
    foreignKey: {
      name: "giver",
      allowNull: false,
    },
  };
  Iou.belongsTo(User, GiverForeignKey);
  User.hasMany(Iou, GiverForeignKey);

  const ReceiverForeignKey = {
    foreignKey: {
      name: "receiver",
      allowNull: true,
    },
  };
  Iou.belongsTo(User, ReceiverForeignKey);
  User.hasMany(Iou, ReceiverForeignKey);

  const ParentOfferForeignKey = {
    foreignKey: {
      name: "parent_offer",
      allowNull: true,
    },
  };
  Iou.belongsTo(Offer, ParentOfferForeignKey);
  Offer.hasMany(Iou, ParentOfferForeignKey);
}

export default Iou;
