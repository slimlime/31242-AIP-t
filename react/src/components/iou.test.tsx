import React from "react";
import { shallow } from "enzyme";
import Iou from "./iou";

describe("<Iou />", () => {
  it("should render correctly", () => {
    const wrapper = shallow(
      <Iou
        iou={{
          id: "1",
          item: { id: "201", display_name: "Coffee" },
          giver: { username: "James", display_name: "James Lee" },
          receiver: { username: "Kevin", display_name: "Kevin Lueng" },
          parent_request: null,
          proof_of_debt: "Some ID",
          proof_of_completion: "Some ID",
          created_time: "02/02/2020",
          claimed_time: "02/02/2020",
          is_claimed: true,
        }}
        iouType={0}
        refreshTable={() => {
          return "substitute test function";
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
