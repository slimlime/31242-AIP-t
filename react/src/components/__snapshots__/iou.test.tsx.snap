// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Iou /> should render correctly 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  spacing={1}
  xs={12}
>
  <WithStyles(ForwardRef(Grid))
    id="requestItemContainer"
    item={true}
    xs={7}
  >
    <IouFavourSingle
      giverDisplayName="James"
      item={
        Object {
          "display_name": "Coffee",
          "id": "201",
        }
      }
      recieverDisplayName="Kevin"
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    id="requestItemContainer"
    item={true}
    xs={2}
  >
    <IouProof
      imagePK="Some ID"
      proof_of_debt={true}
    />
    <IouProof
      imagePK="Some ID"
      proof_of_debt={false}
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    id="requestProofContainer"
    item={true}
    xs={3}
  >
    <IouComplete
      author="Kevin"
      claimed_time="02/02/2020"
      completed_by="James"
      created_time="02/02/2020"
      details=""
      id="1"
      iouType={0}
      is_completed={true}
      refreshTable={[Function]}
      rewards={
        Array [
          Object {
            "giver": Object {
              "display_name": "James",
              "username": "James",
            },
            "id": "",
            "item": Object {
              "display_name": "Coffee",
              "id": "201",
            },
          },
        ]
      }
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
